File: 309-best-time-to-buy-and-sell-stock-with-cooldown/best-time-to-buy-and-sell-stock-with-cooldown.cpp
class Solution {
public:
    int recur(vector<int> &prices, int i, int buy, vector<vector<int>> &dp){
        int n = prices.size();
        if(i >= n) return 0;
        if(dp[i][buy] != -1) return dp[i][buy];
        if(buy){
            return dp[i][buy] = max(-prices[i] + recur(prices, i+1, 0, dp), recur(prices, i+1, 1, dp));
        } else {
            return dp[i][buy] = max(prices[i] + recur(prices, i+2, 1, dp), recur(prices, i+1, 0, dp));
        }
    }
    int maxProfit(vector<int>& prices) {
        int n = prices.size();
        vector<vector<int>> dp(n, vector<int>(2, -1));
        return recur(prices, 0, 1, dp);
    }
};

File: 309-best-time-to-buy-and-sell-stock-with-cooldown/best-time-to-buy-and-sell-stock-with-cooldown.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        dp = [[-1] * 2 for _ in range(len(prices)+1)]

        def recur(i, buy):
            if i >= len(prices):
                return 0

            if dp[i][buy] != -1:
                return dp[i][buy]
            
            ans = 0
            if buy:
                ans = max(-prices[i] + recur(i+1, 0), 0 + recur(i+1, 1))
            else:
                ans = max(prices[i] + recur(i+2, 1), 0 + recur(i+1, 0))
            
            dp[i][buy] = ans

            return ans
        
        result = recur(0, 1)
        return result
        